src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/IconUrlToImageCacheConverter.cs

```diff
@@ -54,8 +61,62 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
 
             var iconBitmapImage = new BitmapImage();
             iconBitmapImage.BeginInit();
-            iconBitmapImage.UriSource = iconUrl;
 
+            BitmapSource imageResult;
+
+            if (values.Length > 1
+                && values[1] != null
+                && values[1] != System.Windows.DependencyProperty.UnsetValue
+                && iconUrl.IsFile
+                && !string.IsNullOrEmpty(iconUrl.Fragment)
+                && iconUrl.Fragment.Length > 1) // is it an embedded icon?
+            {
+                try
+                {
+                    var par = values[1] as PackageArchiveReader;
+                    var iconEntry = Uri.UnescapeDataString(iconUrl.Fragment).Substring(1);
+                    iconBitmapImage.StreamSource = par.GetEntry(iconEntry).Open(); // This stream is closed in BitmapImage events
+
+                    iconBitmapImage.DecodeFailed += (sender, args) =>
```

Comment:
My assumption is there's no need to unsubscribe from these events. If so... the delegates are OK.

Reply:
The only time when the iconBitmapImage is used after it is returned by the converter is in the events itself. Here, I wanted to capture the PackageArchiveMetadata instance so I can close it _after_ all image processing is done. Thus, there's no need to unsubscribe.